File: T03/query.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define PAGE_SIZE 400

// Definição da estrutura do registro de aluno
struct Aluno {
    int seq_aluno;
    char codigo_curso[4];
    char nome_aluno[34];
};

// Definição da estrutura do registro de curso
struct Curso {
    int seq_curso;
    char codigo_curso[4];
    char nome_curso[18];
};

// Função para ler uma página do arquivo de alunos
int readAlunoPage(FILE* file, struct Aluno* page, int pageNumber) {
    fseek(file, pageNumber * PAGE_SIZE, SEEK_SET);
    return fread(page, sizeof(struct Aluno), PAGE_SIZE / sizeof(struct Aluno), file);
}

// Função para ler uma página do arquivo de cursos
int readCursoPage(FILE* file, struct Curso* page, int pageNumber) {
    fseek(file, pageNumber * PAGE_SIZE, SEEK_SET);
    return fread(page, sizeof(struct Curso), PAGE_SIZE / sizeof(struct Curso), file);
}

// Função para realizar a junção dos registros
void mergeJoin(FILE* alunoFile, FILE* cursoFile, int bufferPoolSize, int m, int n) {
    struct Aluno alunoPage[PAGE_SIZE / sizeof(struct Aluno)];
    struct Curso cursoPage[PAGE_SIZE / sizeof(struct Curso)];

    int alunoPageNumber = 0;
    int cursoPageNumber = 0;

    // Buffer Pool para alunos e cursos
    struct Aluno* alunoBufferPool = (struct Aluno*)malloc(PAGE_SIZE * bufferPoolSize);
    struct Curso* cursoBufferPool = (struct Curso*)malloc(PAGE_SIZE * bufferPoolSize);
    int alunoBufferIndex = 0;
    int cursoBufferIndex = 0;

    int alunoBufferFilled = readAlunoPage(alunoFile, alunoBufferPool, alunoPageNumber);
    alunoPageNumber++;
    int cursoBufferFilled = readCursoPage(cursoFile, cursoBufferPool, cursoPageNumber);
    cursoPageNumber++;

    while (alunoBufferFilled > 0) {
        int alunoIndex = 0;
        int cursoIndex = 0;

        while (alunoIndex < alunoBufferFilled && cursoIndex < cursoBufferFilled) {
            if (strcmp(alunoBufferPool[alunoIndex].codigo_curso, cursoBufferPool[cursoIndex].codigo_curso) < 0) {
                alunoIndex++;
            } else if (strcmp(alunoBufferPool[alunoIndex].codigo_curso, cursoBufferPool[cursoIndex].codigo_curso) > 0) {
                cursoIndex++;
            } else {
                // Encontrou uma correspondência, imprime os dados
                printf("SEQ_ALUNO: %d\n", alunoBufferPool[alunoIndex].seq_aluno);
                printf("CODIGO_CURSO: %s\n", alunoBufferPool[alunoIndex].codigo_curso);
                printf("NOME_ALUNO: %s\n\n", alunoBufferPool[alunoIndex].nome_aluno);
                
                // Avança para o próximo registro do buffer de alunos
                alunoIndex++;
            }
        }

        if (alunoIndex == alunoBufferFilled) {
            // Buffer de alunos totalmente percorrido, carrega a próxima página
            alunoBufferFilled = readAlunoPage(alunoFile, alunoBufferPool, alunoPageNumber);
            alunoPageNumber++;
        }

        if(alunoPageNumber>=m)return;

        if (cursoIndex == cursoBufferFilled) {
            // Buffer de cursos totalmente percorrido, carrega a próxima página
            cursoBufferFilled = readCursoPage(cursoFile, cursoBufferPool, cursoPageNumber);
            cursoPageNumber++;
        }

        if(cursoPageNumber>=n)return;
    }

    free(alunoBufferPool);
    free(cursoBufferPool);
}

int main() {
    int n,m;
    printf("Digite o tamanho (em paginas) do arquivo de alunos (N): ");
    scanf("%d",&n);
    printf("Digite o tamanho (em paginas) do arquivo de cursos (M): ");
    scanf("%d",&m);

    FILE* alunoFile = fopen("aluno.data", "rb");
    FILE* cursoFile = fopen("curso.data", "rb");

    if (!alunoFile || !cursoFile) {
        printf("Erro ao abrir os arquivos de entrada.\n");
        return 1;
    }

    int bufferPoolSize;
    printf("Digite o tamanho do buffer pool (B): ");
    scanf("%d", &bufferPoolSize);

    // Realiza a junção dos registros
    mergeJoin(alunoFile, cursoFile, bufferPoolSize,m,n);

    fclose(alunoFile);
    fclose(cursoFile);

    return 0;
}

File: T03/query.py
def merge_sort_join_file(aluno_file, curso_file, buffer_pool_size):
    PAGE_SIZE = 400  # Tamanho de uma página em bytes

    # Função para ler uma página de registros de um arquivo
    def read_page(file, page_num):
        file.seek(page_num * PAGE_SIZE)
        page_data = file.read(PAGE_SIZE)
        records = page_data.strip().split('\n')
        return records

    # Função para realizar a junção das páginas ordenadas
    def merge_pages(aluno_page, curso_page):
        result = []
        i = j = 0
        while i < len(aluno_page) and j < len(curso_page):
            aluno_data = aluno_page[i].split(',')
            curso_data = curso_page[j].split(',')
            codigo_aluno = aluno_data[1]
            codigo_curso = curso_data[1]

            if codigo_aluno == codigo_curso:
                # Realizar a combinação das linhas de aluno e curso
                row = {
                    'SEQ_ALUNO': int(aluno_data[0]),
                    'CODIGO_CURSO': codigo_aluno,
                    'NOME_ALUNO': aluno_data[2],
                    'NOME_CURSO': curso_data[2]
                }
                result.append(row)
                i += 1
                j += 1
            elif codigo_aluno < codigo_curso:
                i += 1
            else:
                j += 1

        return result

    # Abrir os arquivos de aluno e curso
    with open(aluno_file, 'rb') as aluno_f, open(curso_file, 'rb') as curso_f:
        # Definir o tamanho do buffer pool em número de páginas
        buffer_size = buffer_pool_size

        # Inicializar o buffer pool
        buffer_pool = [None] * buffer_size

        # Ler as primeiras páginas de cada arquivo
        aluno_page_num = 0
        curso_page_num = 0
        buffer_index = 0
        while aluno_page_num < buffer_size and curso_page_num < buffer_size:
            buffer_pool[buffer_index] = (
                read_page(aluno_f, aluno_page_num),
                read_page(curso_f, curso_page_num)
            )
            buffer_index += 1
            aluno_page_num += 1
            curso_page_num += 1

        # Realizar a junção das páginas
        result = []
        while True:
            # Encontrar a menor chave de curso no buffer pool
            min_curso_key = float('inf')
            min_curso_index = None
            for i, (aluno_page, curso_page) in enumerate(buffer_pool):
                if curso_page and curso_page[0].split(',')[1] < min_curso_key:
                    min_curso_key = curso_page[0].split(',')[1]
                    min_curso_index = i

            if min_curso_index is None:
                # Não há mais páginas de curso no buffer pool
                break

            # Encontrar as páginas de aluno correspondentes
            matching_aluno_pages = []
            for i, (aluno_page, _) in enumerate(buffer_pool):
                if aluno_page and aluno_page[0].split(',')[1] == min_curso_key:
                    matching_aluno_pages.append(i)

            if not matching_aluno_pages:
                # Não há páginas de aluno correspondentes
                break

            # Realizar a junção das páginas correspondentes
            for aluno_index in matching_aluno_pages:
                aluno_page, curso_page = buffer_pool[aluno_index]
                result.extend(merge_pages(aluno_page, curso_page))

                # Remover as páginas do buffer pool
                buffer_pool[aluno_index] = (None, None)

                # Ler a próxima página de aluno correspondente, se existir
                aluno_page_num += 1
                if aluno_page_num < buffer_size:
                    buffer_pool[aluno_index] = (
                        read_page(aluno_f, aluno_page_num),
                        curso_page
                    )

            # Atualizar as páginas de curso do buffer pool
            curso_page_num += 1
            if curso_page_num < buffer_size:
                buffer_pool[min_curso_index] = (
                    aluno_page,
                    read_page(curso_f, curso_page_num)
                )
            else:
                buffer_pool[min_curso_index] = (aluno_page, None)

    return result

# Exemplo de utilização
aluno_file = 'aluno.dat'
curso_file = 'curso.dat'
buffer_pool_size = 20  # Tamanho do buffer pool em número de páginas

result = merge_sort_join_file(aluno_file, curso_file, buffer_pool_size)
for row in result:
    print(row)
